manual/manual/cmds/instrumented-runtime.etex

```diff
@@ -0,0 +1,148 @@
+
+\begin{verbatim}
+       ocamlopt -runtime-variant i program.ml -o program
+\end{verbatim}
+
+The resulting program can then be traced by running it with the environment
+variable {\tt OCAML\_EVENTLOG\_ENABLED}:
+
+\begin{verbatim}
+        OCAML_EVENTLOG_ENABLED=1 ./program
+\end{verbatim}
+
+The program will execute and generate a trace file in the
+current directory.
+
+\subsection{ss:instr-runtime-prefix}{Trace filename}
+
+The default trace filename is {\tt caml-\{pid\}-eventlog}, where {\tt \{pid\}}
+is the process identifier of the traced program.
+
+This filename can also be specified using the
+{\tt OCAML\_EVENTLOG\_FILE} environment variable.
+The given path will be suffixed with {\tt .pid.eventlog}.
+
+\begin{verbatim}
+        OCAML_EVENTLOG_FILE=/tmp/some/prefix OCAML_EVENTLOG_ENABLED=1 ./program
+\end{verbatim}
+
+In this example, the trace will be available at path
+{\tt /tmp/some/prefix.pid.eventlog}.
+
+\subsection{ss:instr-runtime-pause}{Pausing and resuming tracing}
+Primitives are available to control events collection at runtime.
+
+{\tt OCAML\_EVENTLOG\_ENABLED} can be set to the {\tt p} flag in order
+to {\em pause} events collection after the program has started.
+
+\begin{verbatim}
+        OCAML_EVENTLOG_ENABLED=p ./program
+\end{verbatim}
+
+The program will have to resume events collection explicitly.
+This can be done via {\tt Gc.eventlog_resume} and
+{\tt Gc.eventlog_pause}, available in the {\stdmoduleref{Gc}} module.
+
+\subsection{ss:instr-runtime-limitations}{Limitations}
+
+The instrumented runtime does not support the {\tt fork} system call.
+A child process forked from an instrumented program will not be traced.
+
+The instrumented runtime aims to provide insight into the runtime's execution
+while maintaining a low overhead.
+However, the impact on performance may increase based on the pressure put on the
+garbage collector.
+
+For a program with an extended running time where the collection of only a
+small sample of events is required, using the {\em eventlog_resume} and
+{\em eventlog_pause} primitives may help relieve some of the
+tracing induced performance impact.
+
+\section{s:instr-runtime-read}{Read traces}
+
+Traces generated by the instrumented runtime can be analyzed with tooling
+available outside of the OCaml distribution.
+
+A complete trace is comprised of a {\em metadata file} and a {\em trace file}.
+Two simple ways to work with the traces are the {\em eventlog-tools} and
+{\em babeltrace} libraries.
+
+\subsection{ss:instr-runtime-tools}{eventlog-tools}
+{\em eventlog-tools} is a set of libraries and tools written in OCaml
+that allow to decode and perform basic format conversions and analysis.
+
+For more information about {\em eventlog-tools}, refer to the project's
+main page:
+\ifouthtml
+\ahref{https://github.com/ocaml-multicore/eventlog-tools}
+      {https://github.com/ocaml-multicore/eventlog-tools}.
+\else
+{\tt https://github.com/ocaml-multicore/eventlog-tools}
+\fi
```

Comment:
In addition to pointing at the page for more documentation (which is fine), I would like to see two representative examples of usage of the tools; one to print an histogram (why not show the actual output on an example and say a word about what we see?), and also the command sequence to convert and visualize the trace in chromium. Basically I would expect people reading this manual page to be able to do basic tracing of their OCaml program by just reusing examples shown in the documentation.

Reply:
This sounds good, I will provide such examples shortly.

Reply:
I have one extra question regarding this particular request:
Should the inclusion of such examples be done in the manual page, or it could be done in the eventlog-tools repository itself ? (for the eventlog-tools related  demonstrations.)
As for `babeltrace`, I think it is fine to include a demonstration as well as a small code snippet showing basic decoding of an OCaml trace, for processing. (as is our usecase in the multicore team, using Jupyter.)

Reply:
> Should the inclusion of such examples be done in the manual page, or it could be done in the eventlog-tools repository itself ?

I think it is fine if it is in the eventlog-tools repository, and arguably even better: if you change the tool (and remember to update the documentation), all manual pages that refer to the right repository will get the updated documentation.

I think the documentation review process is mostly useful to ensure that the documentation *exists*. We have more leeway on where precisely the documentation is, as long as you are actually forced to write it :-)